File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>>ans=new ArrayList<>();
        if(root == null) return ans;
        Queue<TreeNode>q=new LinkedList<>();
        q.add(root);
        boolean LeToRe=true;
        while(!q.isEmpty()){
            int s= q.size();
            List<Integer>le=new ArrayList<>();
            for(int i=0;i<s;i++){
                TreeNode t= q.poll();
                le.add(t.val);
                if(t.left != null){
                    q.add(t.left);
                }
                if(t.right !=null){
                    q.add(t.right);
                }
            }
            if(!LeToRe){
                Collections.reverse(le);
            }
            ans.add(le);
            LeToRe= !LeToRe;
        }
        return ans;
    }
}

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>>ans;
        queue<TreeNode*>q;
        if(root == NULL) return ans;
        q.push(root);
        bool LeToRe = true;
        while(!q.empty()){
            int s=q.size();
            vector<int>le;
            for(int i=0;i<s;i++){
                TreeNode* t=q.front();
                q.pop();
                le.push_back(t->val);
                if(t->left){
                    q.push(t->left);
                }
                if(t->right){
                    q.push(t->right);
                }
            }
            if(!LeToRe){
                reverse(le.begin(),le.end());
            }
            ans.push_back(le);
            LeToRe= !LeToRe;
        }
        return ans;
    }
};